# Breast Cancer Tumors: A Story Told By Data

**Breast cancer**, a type of cancer caused by cancerous cells formed in the breast, and majorly occurs in women, these cancerous cells form tumours that can be seen x-rays.

In this article we'll be using the Python **pandas** library to prepare a data set that contains several properties of tumor, cancerous and non-cancerous, recorded as observed from an x-ray, for further analysis.

Let's start by importing the pandas package for wrangling

In [22]:
import pandas as pd

## Reading The Dataset

Next, we'll be reading the csv file that contains our dataset. You can get the dataset from [here](https://www.kaggle.com/baravindkumar/breast-cancer-cancer-non-cancer-classification).

In [5]:
# read the dataset from the csv file
cancer_df = pd.read_csv("breast_cancer.csv")
# view the data
cancer_df

Unnamed: 0  mean radius  mean texture  mean perimeter  mean area  \
0             0        17.99         10.38          122.80     1001.0   
1             1        20.57         17.77          132.90     1326.0   
2             2        19.69         21.25          130.00     1203.0   
3             3        11.42         20.38           77.58      386.1   
4             4        20.29         14.34          135.10     1297.0   
..          ...          ...           ...             ...        ...   
564         564        21.56         22.39          142.00     1479.0   
565         565        20.13         28.25          131.20     1261.0   
566         566        16.60         28.08          108.30      858.1   
567         567        20.60         29.33          140.10     1265.0   
568         568         7.76         24.54           47.92      181.0   

     mean smoothness  mean compactness  mean concavity  mean concave points  \
0            0.11840           0.27760         0.30010              0.14710   
1            0.08474           0.07864         0.08690              0.07017   
2            0.10960           0.15990         0.19740              0.12790   
3            0.14250           0.28390         0.24140              0.10520   
4            0.10030           0.13280         0.19800              0.10430   
..               ...               ...             ...                  ...   
564          0.11100           0.11590         0.24390              0.13890   
565          0.09780           0.10340         0.14400              0.09791   
566          0.08455           0.10230         0.09251              0.05302   
567          0.11780           0.27700         0.35140              0.15200   
568          0.05263           0.04362         0.00000              0.00000   

     mean symmetry  ...  worst texture  worst perimeter  worst area  \
0           0.2419  ...          17.33           184.60      2019.0   
1           0.1812  ...          23.41           158.80      1956.0   
2           0.2069  ...          25.53           152.50      1709.0   
3           0.2597  ...          26.50            98.87       567.7   
4           0.1809  ...          16.67           152.20      1575.0   
..             ...  ...            ...              ...         ...   
564         0.1726  ...          26.40           166.10      2027.0   
565         0.1752  ...          38.25           155.00      1731.0   
566         0.1590  ...          34.12           126.70      1124.0   
567         0.2397  ...          39.42           184.60      1821.0   
568         0.1587  ...          30.37            59.16       268.6   

     worst smoothness  worst compactness  worst concavity  \
0             0.16220            0.66560           0.7119   
1             0.12380            0.18660           0.2416   
2             0.14440            0.42450           0.4504   
3             0.20980            0.86630           0.6869   
4             0.13740            0.20500           0.4000   
..                ...                ...              ...   
564           0.14100            0.21130           0.4107   
565           0.11660            0.19220           0.3215   
566           0.11390            0.30940           0.3403   
567           0.16500            0.86810           0.9387   
568           0.08996            0.06444           0.0000   

     worst concave points  worst symmetry  worst fractal dimension  outcome  
0                  0.2654          0.4601                  0.11890        0  
1                  0.1860          0.2750                  0.08902        0  
2                  0.2430          0.3613                  0.08758        0  
3                  0.2575          0.6638                  0.17300        0  
4                  0.1625          0.2364                  0.07678        0  
..                    ...             ...                      ...      ...  
564                0.2216          0.2060      

Taking a look at the data set, it's obvious that the first column **"Unnamed: 0"** is not needed and it's very similar to the **index column** on the right, therefore the column will be dropped.

In [32]:
cancer_df.drop(labels=["Unnamed: 0"], axis=1)

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0          17.99         10.38          122.80     1001.0          0.11840   
1          20.57         17.77          132.90     1326.0          0.08474   
2          19.69         21.25          130.00     1203.0          0.10960   
3          11.42         20.38           77.58      386.1          0.14250   
4          20.29         14.34          135.10     1297.0          0.10030   
..           ...           ...             ...        ...              ...   
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
0             0.27760         0.30010              0.14710         0.2419   
1             0.07864         0.08690              0.07017         0.1812   
2             0.15990         0.19740              0.12790         0.2069   
3             0.28390         0.24140              0.10520         0.2597   
4             0.13280         0.19800              0.10430         0.1809   
..                ...             ...                  ...            ...   
564           0.11590         0.24390              0.13890         0.1726   
565           0.10340         0.14400              0.09791         0.1752   
566           0.10230         0.09251              0.05302         0.1590   
567           0.27700         0.35140              0.15200         0.2397   
568           0.04362         0.00000              0.00000         0.1587   

     mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                   0.07871  ...          17.33           184.60      2019.0   
1                   0.05667  ...          23.41           158.80      1956.0   
2                   0.05999  ...          25.53           152.50      1709.0   
3                   0.09744  ...          26.50            98.87       567.7   
4                   0.05883  ...          16.67           152.20      1575.0   
..                      ...  ...            ...              ...         ...   
564                 0.05623  ...          26.40           166.10      2027.0   
565                 0.05533  ...          38.25           155.00      1731.0   
566                 0.05648  ...          34.12           126.70      1124.0   
567                 0.07016  ...          39.42           184.60      1821.0   
568                 0.05884  ...          30.37            59.16       268.6   

     worst smoothness  worst compactness  worst concavity  \
0             0.16220            0.66560           0.7119   
1             0.12380            0.18660           0.2416   
2             0.14440            0.42450           0.4504   
3             0.20980            0.86630           0.6869   
4             0.13740            0.20500           0.4000   
..                ...                ...              ...   
564           0.14100            0.21130           0.4107   
565           0.11660            0.19220           0.3215   
566           0.11390            0.30940           0.3403   
567           0.16500            0.86810           0.9387   
568           0.08996            0.06444           0.0000   

     worst concave points  worst symmetry  worst fractal dimension  outcome  
0                  0.2654          0.4601                  0.11890        0  
1                  0.1860          0.2750                  0.08902        0  
2                  0.2430          0.3613                  0.08758        0  
3                  0.2575          0.6638                  0.17300        0  
4                  0.1625          0.2364                  

### More Info on Data
To get a less ambiguous idea of how the dataset looks like: 

In [24]:
# More info on the dataset
cancer_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 32 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Unnamed: 0               569 non-null    int64  
 1   mean radius              569 non-null    float64
 2   mean texture             569 non-null    float64
 3   mean perimeter           569 non-null    float64
 4   mean area                569 non-null    float64
 5   mean smoothness          569 non-null    float64
 6   mean compactness         569 non-null    float64
 7   mean concavity           569 non-null    float64
 8   mean concave points      569 non-null    float64
 9   mean symmetry            569 non-null    float64
 10  mean fractal dimension   569 non-null    float64
 11  radius error             569 non-null    float64
 12  texture error            569 non-null    float64
 13  perimeter error          569 non-null    float64
 14  area error               5

After looking through the data set the column labelled **"outcome"** contains the 2 classes that indicates whether the tumor is cancerous or not with the values 0 and 1


## Removing Null Values
Let's attempt to clean the data by checking if the data contains any null values.

In [11]:
# check the dataframe for any null values
cancer_df.isnull().values.any()

False

Now that it's confirmed that the dataset doesn't contain any null values, we can move on to filtering the data.

## Filtering The Data

- We'll be taking our first step of filtering the data by getting the data for when the tumor actually turned out to be a result of cancerous cells.


In [17]:
# Get the dataframe for when the tumour was caused by cancerours cells
cancerous_df = cancer_df.loc[(cancer_df["outcome"] == 1)]
cancerous_df

Unnamed: 0  mean radius  mean texture  mean perimeter  mean area  \
19           19       13.540         14.36           87.46      566.3   
20           20       13.080         15.71           85.63      520.0   
21           21        9.504         12.44           60.34      273.9   
37           37       13.030         18.42           82.61      523.8   
46           46        8.196         16.84           51.71      201.9   
..          ...          ...           ...             ...        ...   
558         558       14.590         22.68           96.39      657.1   
559         559       11.510         23.93           74.52      403.5   
560         560       14.050         27.15           91.38      600.4   
561         561       11.200         29.37           70.67      386.0   
568         568        7.760         24.54           47.92      181.0   

     mean smoothness  mean compactness  mean concavity  mean concave points  \
19           0.09779           0.08129         0.06664             0.047810   
20           0.10750           0.12700         0.04568             0.031100   
21           0.10240           0.06492         0.02956             0.020760   
37           0.08983           0.03766         0.02562             0.029230   
46           0.08600           0.05943         0.01588             0.005917   
..               ...               ...             ...                  ...   
558          0.08473           0.13300         0.10290             0.037360   
559          0.09261           0.10210         0.11120             0.041050   
560          0.09929           0.11260         0.04462             0.043040   
561          0.07449           0.03558         0.00000             0.000000   
568          0.05263           0.04362         0.00000             0.000000   

     mean symmetry  ...  worst texture  worst perimeter  worst area  \
19          0.1885  ...          19.26            99.70       711.2   
20          0.1967  ...          20.49            96.09       630.5   
21          0.1815  ...          15.66            65.13       314.9   
37          0.1467  ...          22.81            84.46       545.9   
46          0.1769  ...          21.96            57.26       242.2   
..             ...  ...            ...              ...         ...   
558         0.1454  ...          27.27           105.90       733.5   
559         0.1388  ...          37.16            82.28       474.2   
560         0.1537  ...          33.17           100.20       706.7   
561         0.1060  ...          38.30            75.19       439.6   
568         0.1587  ...          30.37            59.16       268.6   

     worst smoothness  worst compactness  worst concavity  \
19            0.14400            0.17730          0.23900   
20            0.13120            0.27760          0.18900   
21            0.13240            0.11480          0.08867   
37            0.09701            0.04619          0.04833   
46            0.12970            0.13570          0.06880   
..                ...                ...              ...   
558           0.10260            0.31710          0.36620   
559           0.12980            0.25170          0.36300   
560           0.12410            0.22640          0.13260   
561           0.09267            0.05494          0.00000   
568           0.08996            0.06444          0.00000   

     worst concave points  worst symmetry  worst fractal dimension  outcome  
19                0.12880          0.2977                  0.07259        1  
20                0.07283          0.3184                  0.08183        1  
21                0.06227          0.2450                  0.07773        1  
37                0.05013          0.1987                  0.06169        1  
46                0.02564          0.3105                  0.07409        1  
..                    ...             ...                      ...      ...  
558               0.11050          0.2258      

- Next, let's try getting a statistical description of the previously filtered data (for when tumors actually turned out to be of cancerous cells) 

In [20]:
# Describe the data when the tumour resulted into cancer.
cancerous_df.describe()

Unnamed: 0  mean radius  mean texture  mean perimeter   mean area  \
count  357.000000   357.000000    357.000000      357.000000  357.000000   
mean   319.899160    12.146524     17.914762       78.075406  462.790196   
std    153.157795     1.780512      3.995125       11.807438  134.287118   
min     19.000000     6.981000      9.710000       43.790000  143.500000   
25%    185.000000    11.080000     15.150000       70.870000  378.200000   
50%    332.000000    12.200000     17.390000       78.180000  458.400000   
75%    453.000000    13.370000     19.760000       86.100000  551.100000   
max    568.000000    17.850000     33.810000      114.600000  992.100000   

       mean smoothness  mean compactness  mean concavity  mean concave points  \
count       357.000000        357.000000      357.000000           357.000000   
mean          0.092478          0.080085        0.046058             0.025717   
std           0.013446          0.033750        0.043442             0.015909   
min           0.052630          0.019380        0.000000             0.000000   
25%           0.083060          0.055620        0.020310             0.015020   
50%           0.090760          0.075290        0.037090             0.023440   
75%           0.100700          0.097550        0.059990             0.032510   
max           0.163400          0.223900        0.410800             0.085340   

       mean symmetry  ...  worst texture  worst perimeter   worst area  \
count     357.000000  ...     357.000000       357.000000   357.000000   
mean        0.174186  ...      23.515070        87.005938   558.899440   
std         0.024807  ...       5.493955        13.527091   163.601424   
min         0.106000  ...      12.020000        50.410000   185.200000   
25%         0.158000  ...      19.580000        78.270000   447.100000   
50%         0.171400  ...      22.820000        86.920000   547.400000   
75%         0.189000  ...      26.510000        96.590000   670.000000   
max         0.274300  ...      41.780000       127.100000  1210.000000   

       worst smoothness  worst compactness  worst concavity  \
count        357.000000         357.000000       357.000000   
mean           0.124959           0.182673         0.166238   
std            0.020013           0.092180         0.140368   
min            0.071170           0.027290         0.000000   
25%            0.110400           0.112000         0.077080   
50%            0.125400           0.169800         0.141200   
75%            0.137600           0.230200         0.221600   
max            0.200600           0.584900         1.252000   

       worst concave points  worst symmetry  worst fractal dimension  outcome  
count            357.000000      357.000000               357.000000    357.0  
mean               0.074444        0.270246                 0.079442      1.0  
std                0.035797        0.041745                 0.013804      0.0  
min                0.000000        0.156600                 0.055210      1.0  
25%                0.051040        0.240600                 0.070090      1.0  
50%                0.074310        0.268700                 0.077120      1.0  
75%                0.097490        0.298300                 0.085410      1.0  
max                0.175000        0.422800                 0.148600      1.0  

[8 rows x 32 columns]

> From the above description, it is very easy to get useful information from the above result, like the average **"mean radius"** for tumors caused by cancerous cells is **12** and there were about **357** cases where the tumors were as a result of cancerous cells, and several other useful descriptions  

- Repeating the same process for cases where the values in the **"outcome"** column is 0, that is the tumor was non-cancerous. 

In [19]:
# Get the dataframe for when the tumour didn't result in cancer
non_cancerous_df = cancer_df.loc[(cancer_df["outcome"]) == 0]
non_cancerous_df

Unnamed: 0  mean radius  mean texture  mean perimeter  mean area  \
0             0        17.99         10.38          122.80     1001.0   
1             1        20.57         17.77          132.90     1326.0   
2             2        19.69         21.25          130.00     1203.0   
3             3        11.42         20.38           77.58      386.1   
4             4        20.29         14.34          135.10     1297.0   
..          ...          ...           ...             ...        ...   
563         563        20.92         25.09          143.00     1347.0   
564         564        21.56         22.39          142.00     1479.0   
565         565        20.13         28.25          131.20     1261.0   
566         566        16.60         28.08          108.30      858.1   
567         567        20.60         29.33          140.10     1265.0   

     mean smoothness  mean compactness  mean concavity  mean concave points  \
0            0.11840           0.27760         0.30010              0.14710   
1            0.08474           0.07864         0.08690              0.07017   
2            0.10960           0.15990         0.19740              0.12790   
3            0.14250           0.28390         0.24140              0.10520   
4            0.10030           0.13280         0.19800              0.10430   
..               ...               ...             ...                  ...   
563          0.10990           0.22360         0.31740              0.14740   
564          0.11100           0.11590         0.24390              0.13890   
565          0.09780           0.10340         0.14400              0.09791   
566          0.08455           0.10230         0.09251              0.05302   
567          0.11780           0.27700         0.35140              0.15200   

     mean symmetry  ...  worst texture  worst perimeter  worst area  \
0           0.2419  ...          17.33           184.60      2019.0   
1           0.1812  ...          23.41           158.80      1956.0   
2           0.2069  ...          25.53           152.50      1709.0   
3           0.2597  ...          26.50            98.87       567.7   
4           0.1809  ...          16.67           152.20      1575.0   
..             ...  ...            ...              ...         ...   
563         0.2149  ...          29.41           179.10      1819.0   
564         0.1726  ...          26.40           166.10      2027.0   
565         0.1752  ...          38.25           155.00      1731.0   
566         0.1590  ...          34.12           126.70      1124.0   
567         0.2397  ...          39.42           184.60      1821.0   

     worst smoothness  worst compactness  worst concavity  \
0              0.1622             0.6656           0.7119   
1              0.1238             0.1866           0.2416   
2              0.1444             0.4245           0.4504   
3              0.2098             0.8663           0.6869   
4              0.1374             0.2050           0.4000   
..                ...                ...              ...   
563            0.1407             0.4186           0.6599   
564            0.1410             0.2113           0.4107   
565            0.1166             0.1922           0.3215   
566            0.1139             0.3094           0.3403   
567            0.1650             0.8681           0.9387   

     worst concave points  worst symmetry  worst fractal dimension  outcome  
0                  0.2654          0.4601                  0.11890        0  
1                  0.1860          0.2750                  0.08902        0  
2                  0.2430          0.3613                  0.08758        0  
3                  0.2575          0.6638                  0.17300        0  
4                  0.1625          0.2364                  0.07678        0  
..                    ...             ...                      ...      ...  
563                0.2542          0.2929      

- Describing the data for when the tumor wasn't cancerous

In [21]:
# Describe the data for when the tumor didn't result to cancer
non_cancerous_df.describe()

Unnamed: 0  mean radius  mean texture  mean perimeter    mean area  \
count  212.000000   212.000000    212.000000      212.000000   212.000000   
mean   223.547170    17.462830     21.604906      115.365377   978.376415   
std    165.308422     3.203971      3.779470       21.854653   367.937978   
min      0.000000    10.950000     10.380000       71.900000   361.600000   
25%     74.500000    15.075000     19.327500       98.745000   705.300000   
50%    202.500000    17.325000     21.460000      114.200000   932.000000   
75%    351.250000    19.590000     23.765000      129.925000  1203.750000   
max    567.000000    28.110000     39.280000      188.500000  2501.000000   

       mean smoothness  mean compactness  mean concavity  mean concave points  \
count       212.000000        212.000000      212.000000           212.000000   
mean          0.102898          0.145188        0.160775             0.087990   
std           0.012608          0.053987        0.075019             0.034374   
min           0.073710          0.046050        0.023980             0.020310   
25%           0.094010          0.109600        0.109525             0.064620   
50%           0.102200          0.132350        0.151350             0.086280   
75%           0.110925          0.172400        0.203050             0.103175   
max           0.144700          0.345400        0.426800             0.201200   

       mean symmetry  ...  worst texture  worst perimeter   worst area  \
count     212.000000  ...     212.000000       212.000000   212.000000   
mean        0.192909  ...      29.318208       141.370330  1422.286321   
std         0.027638  ...       5.434804        29.457055   597.967743   
min         0.130800  ...      16.670000        85.100000   508.100000   
25%         0.174050  ...      25.782500       119.325000   970.300000   
50%         0.189900  ...      28.945000       138.000000  1303.000000   
75%         0.209850  ...      32.690000       159.800000  1712.750000   
max         0.304000  ...      49.540000       251.200000  4254.000000   

       worst smoothness  worst compactness  worst concavity  \
count        212.000000         212.000000       212.000000   
mean           0.144845           0.374824         0.450606   
std            0.021870           0.170372         0.181507   
min            0.088220           0.051310         0.023980   
25%            0.130475           0.244475         0.326425   
50%            0.143450           0.356350         0.404900   
75%            0.155975           0.447850         0.556175   
max            0.222600           1.058000         1.170000   

       worst concave points  worst symmetry  worst fractal dimension  outcome  
count            212.000000      212.000000               212.000000    212.0  
mean               0.182237        0.323468                 0.091530      0.0  
std                0.046308        0.074685                 0.021553      0.0  
min                0.028990        0.156500                 0.055040      0.0  
25%                0.152750        0.276500                 0.076302      0.0  
50%                0.182000        0.310300                 0.087600      0.0  
75%                0.210675        0.359225                 0.102625      0.0  
max                0.291000        0.663800                 0.207500      0.0  

[8 rows x 32 columns]

> From the above description, we can see that there were **212 cases** where the tumor was non-cancerous, and the **mean-radius** of tumors of non-cancerous cells is **17**.

Find the code used in this article [here]()